### DISCUSSION TITLE
Functions to verify if a function/method is viper, asm_thumb or native?

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#10936] [Libraries & Drivers]
Question is basically in the title.

Say I pass a callback to be processed at the C level, how can I identify if the object is callable and a Viper, native or asm_thumb function (IOW, not using bytecodes)?

**Comment by stinos** [comment]
Why would you want to know this? It's a generic function object and it can be called, that's usually all that is needed in C?

**Reply by cve2022** [reply]
Because I want to extend an object and the callback MUST be of one of this types for my extension to work so I want to verify the callback type. My attempts for doing it in MP were unsuccessfull, so I will try doing it in C.

**Reply by cve2022** [reply]
In a discussion  (#9784) @rkompass posted the function below. I would like to know which function gets the types in the comparisons below (on REPL, `type()` returns `<class 'function'>` for both Viper and asm_thumb so there must be something else happening under the hood).


```
def inspect(f, nbytes=200):
    import machine
    import array
    import ubinascii
    @micropython.asm_thumb
    def dummy_asm():
        pass
    @micropython.viper
    def dummy_vip() -> int:
        return 0
    if type(f) == type(dummy_asm):
        offs = 0
    elif type(f) == type(dummy_vip):
        offs = 4
    else:
        raise ValueError('expecting an asm_thumb or viper function')
    baddr = bytes(array.array('O', [f]))
    addr = int.from_bytes(baddr, 'little')
    print('{:s} function object at: 0x{:08x}'.format(('asm_thumb' if offs == 0 else 'viper'), addr))
    print('number of args: %u' % machine.mem32[addr + offs + 4])
    code_addr = machine.mem32[addr + offs + 8]     # 8 for assembler , 12 for viper in uPy 1.19
    print('machine code at: 0x%08x' % code_addr)  # seems to fit even with viper : code begins with 'FE B5' and ends with 'FE BD'
    print('-- code --')
    for i in range(nbytes):
        print('%02x' % machine.mem8[code_addr + i], end='')
    print('\n----------')
```

**Reply by stinos** [reply]
`type` is `mp_obj_get_type` in C (in general when doing things in C you're mostly using things from obj.h)

**Reply by cve2022** [reply]
Thank you, I will take a look.

**Reply by rkompass** [reply]
I found that the `id(type(f))` gives some more detail than just `<class 'function'>`, but it does not distinguish between native and viper.
For a test of this:
```
def py_non():
    pass

def py_ret(x):
    return 42

def py_mul(x):
    return 2*x

@micropython.native
def nat_non():
    pass

@micropython.native
def nat_ret(x):
    return 77

@micropython.native
def nat_add(x):
    return x+1

ar = bytearray(12)
@micropython.viper
def vip_wr():
    a = ptr32(ar)
    a[1] = 1234

@micropython.viper
def vip_ret() -> int:
    return 1

@micropython.viper
def vip_non():
    pass

@micropython.asm_thumb
def asm_shift(r0) -> int:
    mov(r1, r0)
    lsl(r0, r1, 3)        # we can have lsl and lsr with argument of source and constant of shift

@micropython.asm_thumb
def asm_ret():        # ----- asm function that returns 0x40000000
        nop()                  # for alignment
        data(2, 0x4800)        # ldr  r0, [pc, #0]  load r0 with data following the branch
        b(HERE)
        data(4, 0x40000000)    # r0 = 0x40000000  -> SyntaxError: 'data' requires integer arguments; r0 = 0x3fffffff still works
        label(HERE)


def fun_type(f):
    def dummy_py():
        pass
    @micropython.native
    def dummy_nat():
        pass
    @micropython.viper
    def dummy_vip() -> int:
        return 0
    @micropython.asm_thumb
    def dummy_asm():
        pass
    if type(f) == type(dummy_py):
        return 'python'
    elif type(f) == type(dummy_nat):
        return 'native'
    elif type(f) == type(dummy_vip):
        return 'viper'
    elif type(f) == type(dummy_asm):
        return 'assembler'
    else:
        return 'unknown'

print(f'Function: py_non    Type: {fun_type(py_non)} Type_Id: {id(type(py_non))}')
print(f'Function: py_ret    Type: {fun_type(py_ret)} Type_Id: {id(type(py_ret))}')
print(f'Function: py_mul    Type: {fun_type(py_mul)} Type_Id: {id(type(py_mul))}')
print(f'Function: nat_non   Type: {fun_type(nat_non)} Type_Id: {id(type(nat_non))}')
print(f'Function: nat_ret   Type: {fun_type(nat_ret)} Type_Id: {id(type(nat_ret))}')
print(f'Function: nat_add   Type: {fun_type(nat_add)} Type_Id: {id(type(nat_add))}')
print(f'Function: vip_wr    Type: {fun_type(vip_wr)} Type_Id: {id(type(vip_wr))}')
print(f'Function: vip_ret   Type: {fun_type(vip_ret)} Type_Id: {id(type(vip_ret))}')
print(f'Function: vip_non   Type: {fun_type(vip_non)} Type_Id: {id(type(vip_non))}')
print(f'Function: asm_shift Type: {fun_type(asm_shift)} Type_Id: {id(type(asm_shift))}')
print(f'Function: asm_ret   Type: {fun_type(asm_ret)} Type_Id: {id(type(asm_ret))}')
```
It returns:
```
Function: py_non    Type: python Type_Id: 268663424
Function: py_ret    Type: python Type_Id: 268663424
Function: py_mul    Type: python Type_Id: 268663424
Function: nat_non   Type: native Type_Id: 268663460
Function: nat_ret   Type: native Type_Id: 268663460
Function: nat_add   Type: native Type_Id: 268663460
Function: vip_wr    Type: native Type_Id: 268663460
Function: vip_ret   Type: native Type_Id: 268663460
Function: vip_non   Type: native Type_Id: 268663460
Function: asm_shift Type: assembler Type_Id: 268663396
Function: asm_ret   Type: assembler Type_Id: 268663396
```
Perhaps it should distinguish between native and viper though?
Now we have that `type(vip_wr) == type(nat_ret)` returns `True`.
Should this be considered a bug? These variants of type are not documented afaik.
It's worth to have a look at [the types module](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/types/types.py).
Perhaps we should extend this for the @micropython-decorated function variants.

**Reply by cve2022** [reply]
@rkompass , thanks a lot.

I have learned yet another trick, this use of `id` is a neat one. So many things to learn and only 24h in the day. 😃 

> It's worth to have a look at [the types module](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/types/types.py).

I certainly will.

Once again, thanks very much.

**Reply by cve2022** [reply]
On second thought, in my particular case, the distinction between viper, native and asm_thumb is not that important, as long as the callback is in native machine code. In pseudo pseudo code:

```
if not callable(fn) or mp_byte_code(fn):
    raise TypeError ("OOPS! only native code allowed")
else
    install_the_callback(fn)
```
My idea is coding this callback to do the very minimum that requires low latency and then schedule a micropython callback that can do more extensive stuff, but may take longer to start if the scheduler queue is crowded.

**Reply by rkompass** [reply]
If you find some trick to circumvent the python handling of the interrupt, which takes some microseconds, and start the callable code directly (nanoseconds delay) that would be superb.

**Reply by robert-hh** [reply]
If you replace the ISR for a peripheral with your own function, then some time can be saved. It depends on which peripheral you go for. Eventually the previous ISR could be called after your urgent stuff has been done. Then you can get into few µs response time. Some response time using a callback toggling a PIN after PIN IRQ using the standard IRQ handler:

MIMXRT1170  0.75 µs
Teensy 4.1  1.5µs
RP2 Pico  17µs
ESP32: 20 - 350µs (!)

**Reply by cve2022** [reply]
> If you replace the ISR for a peripheral with your own function, then some time can be saved. It depends on which peripheral you go for. Eventually the previous ISR could be called after your urgent stuff has been done. Then you can get into few µs response time. Some response time using a callback toggling a PIN after PIN IRQ using the standard IRQ handler:
> 
> MIMXRT1170 0.75 µs Teensy 4.1 1.5µs RP2 Pico 17µs ESP32: 20 - 350µs (!)

FWIW I am targeting the Timer class. I want to extend it (on RP2 port) to use other alarm interrupt(s). Currently, only the ALARM3 interrupt is being used by the timing subsystem of MP, including Timer objects. The ISR function of the SDK already determines which one of the four ALARM interrupts was triggered, does some housekeeping and calls the handler of the corresponding ALARM.  

My understanding is that at the C level I just need to install the handler(s) for other ALARM interrupt(s) that will simply call the respective callback(s). The callback itself (with native machine code) should do the urgent stuff and, if necessary, call `micropython.schedule()` for other not as urgent tasks.

About the latencies, up to 350us with the ESP32? Wow! :astonished: But in fact, it is not as much the latency as the jitter that concerns me.
